signer/src/network/libp2p/swarm.rs

```diff
@@ -658,4 +682,208 @@ mod tests {
+        assert!(*p2p_peers_1a[0].last_dialed_at - utc_now < time::Duration::seconds(5));
+
+        // Verify that context 2 has the peer from context 1.
+        assert_eq!(p2p_peers_2a.len(), 1);
+        assert_eq!(*p2p_peers_2a[0].peer_id, key1_pub.into());
+        assert_eq!(p2p_peers_2a[0].public_key, key1_pub);
+        assert_eq!(*p2p_peers_2a[0].address, swarm1_addr.clone());
+        assert!(*p2p_peers_2a[0].last_dialed_at - utc_now < time::Duration::seconds(5));
+
+        // Trigger shutdown
+        term1.signal_shutdown();
+        term2.signal_shutdown();
+    }
+
+    #[test_log::test(tokio::test)]
+    async fn libp2p_swarm_bootstraps_with_known_peers() {
+        let mut rng = get_rng();
+        let swarm1_addr = Multiaddr::memory(rng.next_u64());
+        let swarm2_addr = Multiaddr::memory(rng.next_u64());
+
+        // PeerId = 16Uiu2HAm46BSFWYYWzMjhTRDRwXHpDWpQ32iu93nzDwd1F4Tt256
+        let key1 = PrivateKey::from_slice(
+            hex::decode("ab0893ecf683dc188c3fb219dd6489dc304bb5babb8151a41245a70e60cb7258")
+                .unwrap()
+                .as_slice(),
+        )
+        .unwrap();
+        let key1_pub = PublicKey::from_private_key(&key1);
+
+        // PeerId = 16Uiu2HAkuyB8ECXxACm8hzQj4vZ2iWrYMF3xcKNf1oJJ1NuQEMvQ
+        let key2 = PrivateKey::from_slice(
+            hex::decode("0dd4077c8bcec09c803f9ba23a0f5b56eba75769b2d1b96a33b579dbbe5055ce")
+                .unwrap()
+                .as_slice(),
+        )
+        .unwrap();
+        let key2_pub = PublicKey::from_private_key(&key2);
+
+        // Create two contexts with different keys but in same signer set.
+        let context1 = TestContext::builder()
+            .with_in_memory_storage()
+            .with_mocked_clients()
+            .with_private_key(key1)
+            .build();
+
+        let context2 = TestContext::builder()
+            .with_in_memory_storage()
+            .with_mocked_clients()
+            .with_private_key(key2)
+            .build();
+
+        // Add each key to the other's signer set so they can connect.
+        context1.state().current_signer_set().add_signer(key2_pub);
+        context2.state().current_signer_set().add_signer(key1_pub);
+
+        // Create the swarms. Swarm 1 will bootstrap to swarm 2 via known peers.
+        let swarm1 = SignerSwarmBuilder::new(&key1)
+            .enable_mdns(false)
+            .enable_kademlia(false)
+            .enable_autonat(false)
+            .enable_memory_transport(true)
+            .with_initial_bootstrap_delay(Duration::ZERO) // Connect immediately
+            .add_known_peers(&[(key2_pub.into(), swarm2_addr.clone())])
+            .add_listen_endpoint(swarm1_addr.clone())
+            .build()
+            .expect("Failed to build swarm 1");
+        let swarm2 = SignerSwarmBuilder::new(&key2)
+            .enable_mdns(false)
+            .enable_kademlia(false)
+            .enable_autonat(false)
```

Comment:
Do we want to test it with this features enabled? At least we use branching based on if kademlia is on, so I think we should try with kademlia = true ?

Reply:
For this test I'm explicitly not wanting kademlia involved because I want to make sure that the peers are being populated from the database and not some other source.